packages/runner/src/storage/interface.ts

```diff
@@ -88,3 +99,180 @@ export interface IStorageProvider {
    */
   getReplica(): string | undefined;
 }
+
+/**
+ * This is successor to the current `IStorageProvider` which provides a
+ * transactional interface.
+ */
+export interface IStorageProviderV2 {
+  /**
+   * Creates a new transaction that can be used to build up a change set that
+   * can be committed transactionally. It ensures that all reads are consistent
+   * and no affecting changes takes place until the transaction is committed. If
+   * upstream changes are made since transaction is created that updates any of
+   * the read values transaction will fail on commit.
+   */
+  fork(): IStorageTransaction;
+}
+
+/**
+ * Representation of a storage transaction, which can be used to query facts and
+ * assert / retract while maintaining consistency guarantees. Storage ensures
+ * that transactions retain consistent view of the whole storage through it's
+ * lifetime by notifying pending transaction of every change that is integrated
+ * into the storage, if changes affect any data read through a transaction
+ * lifecycle it can not be committed because it would violate consistency. If
+ * no change occurs or changes do not affect any data read it would not affect
+ * transaction consistency guarantees and therefor committing transaction will
+ * be send to the upstream storage provider which will either accept if no
+ * invariants have being invalidated in the meantime or rejected and fail commit.
+ */
+export interface IStorageTransaction {
+  /**
+   * Transaction can be cancelled which causes storage provider to stop keeping
```

Comment:
The interface doesn't have subscribe yet, but once it has we should note here that `abort` doesn't stop the subscriptions. Useful to get a callback for retrying once e.g. data caught up.